scripts/gen-universe.py

```diff
@@ -499,5 +496,78 @@ def create_index_entry(packages):
     return entry
 
 
+def v3_to_v2_package(v3_package):
+    """Converts a v3 package to a v2 package
+
+    :param v3_package: a v3 package
+    :type v3_package: dict
+    :return: a v2 package
+    :rtype: dict
+    """
+    package = copy.deepcopy(v3_package)
+    package.pop('minDcosReleaseVersion', None)
+    package['packagingVersion'] = "2.0"
+
+    resource = package.pop('resource', None)
+    if resource:
+        cli = resource.pop('cli', None)
+        if cli and 'command' not in package:
+            print(('WARNING: Removing binary CLI from ({}, {}) without a '
+                  'Python CLI').format(package['name'], package['version']))
+        package['resource'] = resource
+
+    return package
+
+
+def v4_to_v3_package(v4_package):
+    """Converts a v4 package to a v3 package
+
+    :param v4_package: a v3 package
+    :type v4_package: dict
+    :return: a v3 package
+    :rtype: dict
+    """
+    package = copy.deepcopy(v4_package)
+    package.pop('upgradesFrom', None)
+    package.pop('downgradesTo', None)
+    return package
+
+
+def downgrade_package_to_v2(package):
```

Comment:
Instead of explicit downgrade functions, what if you had something like this?

```python
convert_from = {'4.0': v4_to_v3_package, '3.0': v3_to_v2_package}

def downgrade_package_to(package, target_version):
    # TODO
```

Then implement `downgrade_package_to()` so that it repeatedly looks up the package's version in `convert_from` and applies the associated function, until the package reaches `target_version`.